cpp/cmake_modules/ThirdpartyToolchain.cmake

```
@@ -397,6 +435,69 @@ endif()
 set(THREADS_PREFER_PTHREAD_FLAG ON)
 find_package(Threads REQUIRED)
 
+# ----------------------------------------------------------------------
+# Google double-conversion
+
+macro(build_double_conversion)
+  set(DOUBLE_CONVERSION_PREFIX "${CMAKE_CURRENT_BINARY_DIR}/double-conversion_ep/src/double-conversion_ep")
+  set(DOUBLE_CONVERSION_HOME "${DOUBLE_CONVERSION_PREFIX}")
+  set(double-conversion_INCLUDE_DIRS "${DOUBLE_CONVERSION_PREFIX}/include")
+  set(DOUBLE_CONVERSION_STATIC_LIB "${DOUBLE_CONVERSION_PREFIX}/lib/${CMAKE_STATIC_LIBRARY_PREFIX}double-conversion${CMAKE_STATIC_LIBRARY_SUFFIX}")
+
+  set(DOUBLE_CONVERSION_CMAKE_ARGS
+    ${EP_COMMON_CMAKE_ARGS}
+    "-DCMAKE_INSTALL_PREFIX=${DOUBLE_CONVERSION_PREFIX}")
+
+  ExternalProject_Add(double-conversion_ep
+    ${EP_LOG_OPTIONS}
+    INSTALL_DIR ${DOUBLE_CONVERSION_PREFIX}
+    URL ${DOUBLE_CONVERSION_SOURCE_URL}
+    CMAKE_ARGS ${DOUBLE_CONVERSION_CMAKE_ARGS}
+    BUILD_BYPRODUCTS "${DOUBLE_CONVERSION_STATIC_LIB}")
+
+  add_library(double-conversion STATIC IMPORTED)
+  set_target_properties(double-conversion
+    PROPERTIES IMPORTED_LOCATION "${DOUBLE_CONVERSION_STATIC_LIB}")
+  add_dependencies(toolchain double-conversion_ep)
+  add_dependencies(double-conversion double-conversion_ep)
+  set(double-conversion_LIBRARIES double-conversion)
+endmacro()
+
+macro(double_conversion_config)
+  # Map the newer target to the old, simpler setting
+  if (TARGET double-conversion::double-conversion)
+    set(double-conversion_LIBRARIES double-conversion::double-conversion)
+    get_target_property(double-conversion_INCLUDE_DIRS double-conversion::double-conversion INTERFACE_INCLUDE_DIRECTORIES)
+  endif()
+endmacro()
+
+# ----------------------------------------------------------------------
+# Resolve the dependencies
+
+# TODO: Check for correct supported double-conversion
+if (ARROW_DEPENDENCY_SOURCE STREQUAL "AUTO")
+  find_package(double-conversion QUIET)
+  if (NOT ${double-conversion_FOUND})
```

```suggestion
  if(${double-conversion_FOUND})
    double_conversion_config()
  else()
    build_double_conversion()
  endif()
```

will be more readable.